File: 2024/day13.go
package main

import (
	"lukechampine.com/advent/utils"
)

var input = utils.Input(2024, 13)

type machine struct {
	a, b  utils.Pos
	prize utils.Pos
}

func (m machine) fracs() (n0 int, d0 int, n1 int, d1 int) {
	n := m.a.X*m.prize.Y - m.a.Y*m.prize.X
	d := m.a.X*m.b.Y - m.a.Y*m.b.X
	return (m.prize.Y - m.b.Y*n/d), m.a.Y, n, d
}

func (m machine) winnable() bool {
	n0, d0, n1, d1 := m.fracs()
	return n0%d0 == 0 && n1%d1 == 0
}

func (m machine) winCost() int {
	n0, d0, n1, d1 := m.fracs()
	return 3*n0/d0 + n1/d1
}

func main() {
	var machines []machine
	for _, group := range utils.Split(input, "\n\n") {
		ints := utils.ExtractInts(group)
		machines = append(machines, machine{
			a:     utils.Pos{X: ints[0], Y: ints[1]},
			b:     utils.Pos{X: ints[2], Y: ints[3]},
			prize: utils.Pos{X: ints[4], Y: ints[5]},
		})
	}
	var sum int
	for _, m := range machines {
		if m.winnable() {
			sum += m.winCost()
		}
	}
	utils.Println(sum)

	sum = 0
	for _, m := range machines {
		m.prize.X += 10000000000000
		m.prize.Y += 10000000000000
		if m.winnable() {
			sum += m.winCost()
		}
	}
	utils.Println(sum)
}

File: 2024/rust/src/day13.rs
use crate::utils;

struct Machine {
    a: utils::Point,
    b: utils::Point,
    prize: utils::Point,
}

impl Machine {
    fn from_string(s: &str) -> Self {
        let ints: Vec<_> = utils::ints(s).collect();
        Machine {
            a: (ints[0], ints[1]).into(),
            b: (ints[2], ints[3]).into(),
            prize: (ints[4], ints[5]).into(),
        }
    }

    fn fracs(&self) -> (isize, isize, isize, isize) {
        let n = self.a.x * self.prize.y - self.a.y * self.prize.x;
        let d = self.a.x * self.b.y - self.a.y * self.b.x;
        ((self.prize.y - self.b.y * n / d), self.a.y, n, d)
    }

    fn is_winnable(&self) -> bool {
        let (n0, d0, n1, d1) = self.fracs();
        n0 % d0 == 0 && n1 % d1 == 0
    }

    fn win_cost(&self) -> isize {
        let (n0, d0, n1, d1) = self.fracs();
        3 * n0 / d0 + n1 / d1
    }
}

pub fn part1(input: &str) -> String {
    input
        .split("\n\n")
        .map(|group| Machine::from_string(group))
        .filter_map(|m| m.is_winnable().then_some(m.win_cost()))
        .sum::<isize>()
        .to_string()
}

pub fn part2(input: &str) -> String {
    input
        .split("\n\n")
        .map(|group| Machine::from_string(group))
        .map(|m| Machine {
            prize: m.prize.add((1e13 as i64, 1e13 as i64).into()),
            ..m
        })
        .filter_map(|m| m.is_winnable().then_some(m.win_cost()))
        .sum::<isize>()
        .to_string()
}
